konflux: acs style snapshot

tommartensen opened the pull request:
### Description

<!--
A detailed explanation of the changes in your PR. Feel free to remove this
section if the title of your PR is sufficiently descriptive. To learn more
about contributing to this project, check "*.md" files under:
    https://github.com/stackrox/stackrox/tree/master/.github
-->

change me!

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

<!--
General Availability requirements: https://github.com/stackrox/stackrox/blob/master/PR_GA.md
Feature Flags usage: https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md
-->

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

<!--
If no tests have been contributed, please explain why unless it's obvious,
e.g., the PR is a one-line comment change.
-->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

<!--
Use this space to explain **how you validated** that **your change functions
exactly how you expect it**. Feel free to attach JSON snippets, curl commands,
screenshots, etc. Apply a simple benchmark: would the information you provided
convince any reviewer or any external reader that you did enough to validate
your change.

It is acceptable to assume trust and keep this section light, e.g. as a
bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a
markdown or code comment change only. It is also acceptable to skip testing for
changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting,
fixing, etc. Make sure you validate the change ASAP after it gets merged or
explain in PR when the validation will be performed. Explain here why you
skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation
activities you did manually and why so.
-->

change me!

tommartensen commented:
/retest  main-on-push

tommartensen commented:
/test main-on-push

tommartensen commented:
/test operator-bundle-on-push

tommartensen commented:
/test main-on-push

tommartensen commented:
/test main-on-push

tommartensen commented:
/test operator-bundle-on-push

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,72 @@
+    }
+
+
+def determine_component_name_suffix(application):
+    return application.lstrip("acs-")
```
Is the idea to take the version suffix like `-4-6` from `acs-4-6` and apply it to components? I think, it may need some better description (either in comments or in function name, or both) for the future generations. For the function name, I can suggest `determine_version_suffix` or `get_version_suffix`.

Also, extracting optional suffix may look more self-explanatory than stripping prefix. Yes, that'll likely require a, regular expression but I think the clarity may worth it.

tommartensen replied:
Done in https://github.com/stackrox/stackrox/pull/13577/commits/7f3b94edc43cc0fb3fb17fd9d08d86f555d466f7

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,72 @@
+    application,
+    components
+):
+    snapshot_name = f"{snapshot_name_prefix}-{int(time.time())}"
```
1. While epoch time is ok-ish, it's not human readable and it can be hard to compare which of two snapshots is more recent since the numbers are big. I suggest using iso8601-ish date here, e.g. `20241218T193201Z` or `20241218_193201z`. We don't need to parse that programmatically so can be flexible but more human-friendly than with the epoch time.
2. WDYT about including the result of the `determine-main-image-tag` task in the snapshot name prefix (i.e. passing that in the pipeline)? I think it will be handy at the time of release to see that yes this is Snapshot of the tagged build (of course, tags recognition by the task is subject of another ticket - ROX-26026).

tommartensen replied:
Done in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,72 @@
+            "labels": {
+                "appstudio.openshift.io/build-pipelinerun": pipeline_run_name
+            }
+        },
```
Does it work correctly without providing the explicit `namespace`?

tommartensen replied:
Yes. The SA only knows about the `rh-acs-tenant` namespace.

msugakov replied:
Magic. I feel that we may guard from some failures in the future if we wire up the actual namespace to be provided there directly. OTOH, I realize you'd need to do some extra setup for that so I'd leave it up to you to decide.

tommartensen replied:
Done in https://github.com/stackrox/stackrox/pull/13577/commits/9c3bbbb39d43450ad7679b348ee1c17e7117d973

It takes the namespace of the pipeline run - ie if we ever moved to a different namespace, it would update automatically.

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -658,3 +661,92 @@ spec:
+          },
+          {
+            "component": "collector",
+            "ref": "$(params.collector-full-image-build-repo)@$(tasks.wait-for-collector-full-image.results.IMAGE_DIGEST)",
+            "repository": "$(tasks.wait-for-collector-full-image.results.GIT_REPO)",
+            "revision": "$(tasks.wait-for-collector-full-image.results.GIT_REF)"
```
[nit] There's a bit of naming clash between `"ref"` and `"revision"` that gets assigned to `...GIT_REF`. You could rename `"ref"` to `"image_ref"` to prevent that.

Have you considered creating this JSON in the format of Snapshot? I.e.

```json
            {
                "name": "...",
                "containerImage": "...",
                "source": {
                    "git": {
                        "revision": "...",
                        "url": "..."
                    }
                }
            },
```

Yes, it's a bit longer here but fewer processing in Python (only manipulate Container `name`).

msugakov replied:
As discussed, renaming would be nice but changing structure isn't required.

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -658,3 +661,92 @@ spec:
+          },
+          {
+            "component": "operator-bundle",
+            "ref": "$(tasks.build-container.results.IMAGE_URL)@$(tasks.build-container.results.IMAGE_DIGEST)",
```
`$(tasks.build-container.results.IMAGE_URL)` will include tag. You may want to use `output-image-repo` for the `operator-bundle` for consistency with others, or on the contrary, add tags to others. Or not do anything. Up to you.

tommartensen replied:
Done `output-image-repo` in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -658,3 +661,92 @@ spec:
+            "ref": "$(params.scanner-v4-db-image-build-repo)@$(tasks.wait-for-scanner-v4-db-image.results.IMAGE_DIGEST)",
+            "repository": "$(params.git-url)",
+            "revision": "$(params.revision)"
+          }
```
`roxctl` is missing. We need to release its container too.

tommartensen replied:
Done in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef

msugakov reviewed (commented):
Made the first pass. Please expect more comments in the following iterations.

tommartensen commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,52 @@
+apiVersion: tekton.dev/v1
```
> I did not have companion scripts in the existing tasks so did not have issues around that.

Is this not a companion script? 😉 
https://github.com/stackrox/konflux-tasks/blob/main/tasks/fetch-scanner-v2-data-task.yaml#L57

There is a way to run Python (and other scripting languages) in steps: https://tekton.dev/vault/pipelines-v0.53.x-lts/tasks/#running-scripts-within-steps. 

I would suggest to park this discussion and move the Python script into the Task definition once all other issues are resolved.

msugakov replied:
> Is this not a companion script?

Right, that's the only one like that and I did not move it. Although it's possible and probably makes more sense than keeping it in the repo.

> There is a way to run Python (and other scripting languages) in steps: [tekton.dev/vault/pipelines-v0.53.x-lts/tasks#running-scripts-within-steps](https://tekton.dev/vault/pipelines-v0.53.x-lts/tasks/#running-scripts-within-steps).

Certainly, shebang line is to our service.

> I would suggest to park this discussion and move the Python script into the Task definition once all other issues are resolved.

I'm ok to delay it.

tommartensen commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -658,3 +661,92 @@ spec:
       resolver: bundles
+
+  - name: create-acs-style-snapshot
+    # Explicitly running after all other tasks to ensure that
+    # - there are no failures that should prevent a release of the operator-bundle image (missing RPMs signatures, deprecated base images, ...)
+    # - the source image is present as it is required by EC
+    runAfter:
+    - apply-tags
+    - build-source-image
+    - clair-scan
+    - clamav-scan
+    - deprecated-base-image-check
+    - rpms-signature-scan
+    - sast-snyk-check
+    - push-dockerfile
```
> It'll be a pain to maintain this list... Do you have any ideas how can we keep it up to date as the time passes and tasks get added or changed?

No, I will think about it. A non-technical solution (ie a comment at the top of the file: "if you add tasks, ... update the runAfter" might suffice.

> One option to try would be to just move create-acs-style-snapshot to finally:. What I wonder is how the failure of it would look like and whether we can still receive Slack messages for that.

all `finally` tasks run in parallel, ie we would not get a Slack notification if the snapshot creation failed. For example, if the pipeline times out waiting for an image, it cannot create a snapshot and would fail this task. I don't think we should move it to `finally`.

msugakov replied:
> all finally tasks run in parallel

Bummer

> A non-technical solution (ie a comment at the top of the file: "if you add tasks, ... update the runAfter" might suffice.

This will fall apart rather quickly because most of times engineer scroll directly to the place they need to change without bothering to re-read comments at the top each time.

One option I see is to create a validation script that would use `yq` to query task names and check that all are listed in `runAfter` and put this script to run in CI.

Alternatively, I'd suggest to sacrifice Slack notification for failing Snapshot creation and move the Snapshot creation to `finally`. Still, a failure there would fail the pipeline so we should notice.

Can't see other options.

tommartensen replied:
> One option I see is to create a validation script that would use yq to query task names and check that **all** are listed in runAfter and put this script to run in CI.

By "all", do you mean each and every task under `.spec.tasks` except `create-acs-style-snapshot`? If you mean just the tasks currently listed in the `runAfter`, we would re-implement a pipeline parser as we'd need to build the pipeline graph and understand which tasks are on which level.

tommartensen replied:
A suggestion done in https://github.com/stackrox/stackrox/pull/13577/commits/a245624c223f482d05e3268e872fc647d2dcc436.

tommartensen replied:
Proof: https://github.com/stackrox/stackrox/actions/runs/12415215130/job/34661075839?pr=13577

tommartensen replied:
<img width="916" alt="Screenshot 2024-12-20 at 11 35 29" src="https://github.com/user-attachments/assets/b5ede07b-430e-4a81-8016-60ffba383668" />

Done what we discussed in https://github.com/stackrox/stackrox/pull/13577/commits/a10584c9626485c342a550a482ebe19ddc0b848c

msugakov commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -10,6 +10,10 @@ spec:
   results:
+  - name: GIT_REF
+    description: Git commit for the image's source code if provided as `vcs-ref` image label.
+  - name: GIT_REPO
+    description: Git repository URL for the image's source code if provided as `source-location` image label.
```
I keep thinking about this and I keep not liking it. The thing is: unlike `vcs-ref` that's provided by Konflux automatically, `source-location` is our stuff and it looks not tied to anything in the Dockerfiles. I imagine in a future some engineer may ask themselves "what is this doing here?" and kill the label. Either we need to guard the labels in comments or what may be much simpler is hardcode git repos in the pipeline. We already have params for quay repos, adding git repos wouldn't make it much worse.

WDYT?

tommartensen replied:
I want to avoid hardcoding in the pipeline as much as possible. 

Ideal scenario: We can guard in the script against git revision and repository (LABEL) being empty strings and fail the pipeline. 
If someone would remove the LABEL, we'd notice in the `operator-bundle` failure. 

That would certainly work for the monorepo components and PRs that update the SCANNER/COLLECTOR version - if all engineers would follow up on Konflux build pipeline faillures.

Do you have an explanation why `vcs-url` is not a label required by EC?
https://github.com/release-engineering/rhtap-ec-policy/blob/main/data/rule_data.yml#L76C1-L83C46

msugakov replied:
> Do you have an explanation why `vcs-url` is not a label required by EC?

No idea. Some other labels rules seem quite random to me.

>  If someone would remove the LABEL, we'd notice in the operator-bundle failure.

Yes, that'll be helpful but not ideal. Folks removing the label in Collector or Scanner repos won't see the effect of this.
I guess, let's just do it (if you like it more than hardcoding repos). I'd also ask you to put comments in each konflux.*Dockerfile next to the `source-location` label saying it must not be removed or renamed because it's used for snapshot creation.

tommartensen replied:
Validating the component object in the Python script: https://github.com/stackrox/stackrox/pull/13577/commits/d9202fa536edc937e4795b8960af8627617b8559.

Created https://issues.redhat.com/browse/ROX-27485 to add the comment to the Collector and Scanner v2 Dockerfiles. We don't use the source-location in the components built from this repository, so we could remove the label without breaking the snapshot creation.

msugakov replied:
Continuing here: https://github.com/stackrox/stackrox/pull/13577#discussion_r1894304576

tommartensen commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,52 @@
+        value: $(params.IMAGE_REFS)
+    script: |
+      #!/usr/bin/env bash
+
```
Done in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef

tommartensen commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -36,3 +40,17 @@ spec:
+        "docker://$(params.IMAGE)")"
+      echo -n "$git_ref" | tee "$(results.GIT_REF.path)"
+
+      git_repo="$(skopeo inspect \
```
Done in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef and following commit. 

`\n` is not an allowed delimiter, so chose space instead. There shouldn't be a space in the digest, revision or URL.

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,72 @@
+
+if __name__ == '__main__':
+    application = os.getenv("APPLICATION", "")
+    pipeline_run_name = os.getenv("PIPELINE_RUN_NAME", "")
```
Because the preferred solution is not to have a separate Python script (https://github.com/stackrox/stackrox/pull/13577#discussion_r1891808529), how about the following:

* We use `os.environ` and fail if the env var is missing
* We only access environment variables in the main function to read them and store their values in local variables.

msugakov replied:
Could work but there are `args:` as a way to pass command-line arguments to the script. `args` does the same as passing options on the command line. Just feed it an array. See here https://github.com/stackrox/konflux-tasks/blob/b13ddf4691d9458fd1d9265c94a42cf4d9571068/tasks/fetch-scanner-v2-data-task.yaml#L51-L57

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,72 @@
+
+
+def load_image_refs():
+    return json.loads(os.getenv("IMAGE_REFS", '[]'))
```
done in https://github.com/stackrox/stackrox/pull/13577/commits/aa031ef6d3e4cedc0d2ab3e8016caa113d3de7ef

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,92 @@
+def validate_component(component):
+    assert (
+        component["component"] != ""
+        or component["ref"] != ""
+        or component["revision"] != ""
+        or component["repository"] != ""
```
This must be `and`, not `or`.

```
>>> assert(True or False), "blah!"
>>> assert(True and False), "blah!"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: blah!
>>> 
```

tommartensen replied:
fixed in https://github.com/stackrox/stackrox/pull/13577/commits/c87fd5fc0f26269f5791755e6a9832aa19369f65.
Good catch - I didn't invert the condition when switching from an `if`.

msugakov commented on scripts/check-konflux-pipelines.sh:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
1. Please unify the script.
2. Please make sure the diff is printed so that the developer knows what to put in `runAfter`.

tommartensen replied:
See https://github.com/stackrox/stackrox/pull/13577#discussion_r1893773899

tommartensen commented:
/test main-on-push

tommartensen commented:
/test scanner-v4-db-on-push

tommartensen commented:
@msugakov I have addressed all comments except the ones regarding moving the script to the task definition.
Here is a proof run of the operator-bundle: https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/pipelineruns/operator-bundle-on-push-vmkqw. 

(The remaining components need to be re-added of course)

Remaining tasks:

- move the script into the task definition
  -  Some of the env vars we pass cannot be provided as args (APPLICATION, NAMESPACE, PIPELINE_RUN_NAME) as we read them from the PipelineRun object in the cluster with fieldRefs. I would put the other three into args to the script. 
  - Ensure that all required vars are set (ie no empty strings)
- Re-enable full operator-bundle pipeline
- Add all components to IMAGE_REFS.

msugakov commented:
/retest operator-bundle-on-push

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,103 @@
+    return ""
+
+
+def process_component(component, name_suffix):
+    validate_component(component)
+    if name_suffix != "":
+        name = f"{component['name']}{name_suffix}"
+    else:
+        name = component["name"]
```
[nits]
1. Please rename `name_suffix` -> `version_suffix` for consistency.
2. If the suffix is empty string, you can safely append it, i.e. no need in `if/else`.

```suggestion
def process_component(component, version_suffix):
    validate_component(component)
    name = f"{component['name']}{version_suffix}"
```

3. With that, you can just inline the f-string in the "name" assignment below.

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,103 @@
+    application,
+    components
+):
+    timestamp = datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%dT%H%M%SZ")
+    snapshot_name = f"{snapshot_name_prefix}-{snapshot_version_suffix}-{timestamp}".lower()
```
I think, the code will become cleaner if you create a dedicated function which creates the snapshot name, call it from "main" and only pass the final `snapshot_name` into `construct_snapshot()`.

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,103 @@
+        and component["containerImage"] != ""
+        and component["revision"] != ""
+        and component["repository"] != ""
+    ), "Component must have component name, ref, revision and repository set. Check container image labels."
```
I feel like the recommendation to check container labels is unfortunately very much misplaced here.
This could be countered by adding information about which exact labels are to check for which image but that's even worse.

Imagine this task is moved to a https://github.com/stackrox/konflux-tasks repo and so it's abstracted away from how it's used in the pipeline and what the pipeline does to extract that information, yet this task would give the information about what some other task should do.
Spaghetti.

I think, it may be useful to keep this emptiness check here but validating the presence of labels must be a business of the `wait-for-image` task.

msugakov commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -30,9 +34,12 @@ spec:
+      infos="$(skopeo inspect \
         --retry-times 10 \
-        --format '{{.Digest}}' \
+        --format '{{.Digest}} {{ index .Labels "vcs-ref" }} {{ index .Labels "source-location" }}' \
```
As I mentioned in my other comment, this task should be the place to check the presence of `vcs-ref` and `source-location` labels and to show error when they aren't set.

msugakov replied:
This will also mean that `source-location` has to be kept in all `konflux*Dockerfile`-s, not just the Scanner-V2 and Collector ones.

https://github.com/stackrox/stackrox/pull/13577#discussion_r1893601603

tommartensen replied:
> This will also mean that source-location has to be kept in all konflux*Dockerfile-s, not just the Scanner-V2 and Collector ones.

I updated the task description: https://issues.redhat.com/browse/ROX-27485

tommartensen replied:
Check added in https://github.com/stackrox/stackrox/pull/13577/commits/709e0fa713ea1731877488fdaaebad6e032aebb5

msugakov commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -30,9 +34,12 @@ spec:
         "docker://$(params.IMAGE)")"
-      echo -n "$digest" | tee "$(results.IMAGE_DIGEST.path)"
+
+      echo -n $(cut -f1 -d' ' <<< "${infos}") | tee "$(results.IMAGE_DIGEST.path)"
+      echo -n $(cut -f2 -d' ' <<< "${infos}") | tee "$(results.GIT_REF.path)"
+      echo -n $(cut -f3 -d' ' <<< "${infos}") | tee "$(results.GIT_REPO.path)"
```
Please quote substitutions
```suggestion
      echo -n "$(cut -f1 -d' ' <<< "${infos}")" | tee "$(results.IMAGE_DIGEST.path)"
      echo -n "$(cut -f2 -d' ' <<< "${infos}")" | tee "$(results.GIT_REF.path)"
      echo -n "$(cut -f3 -d' ' <<< "${infos}")" | tee "$(results.GIT_REPO.path)"
```

tommartensen replied:
Done in https://github.com/stackrox/stackrox/pull/13577/commits/709e0fa713ea1731877488fdaaebad6e032aebb5

msugakov reviewed (commented):
Some partial iteration

msugakov commented on scripts/ci/jobs/check-konflux-pipelines.sh:
```diff
@@ -0,0 +1,27 @@
+    expected_runafter="$(yq eval '.spec.tasks[] | select(.name != "create-acs-style-snapshot") | .name' .tekton/operator-bundle-pipeline.yaml | sort)"
+    actual_runafter="$(yq eval '.spec.tasks[] | select(.name == "create-acs-style-snapshot") | .runAfter[]' .tekton/operator-bundle-pipeline.yaml)"
+
+    if ! DIFF=$(diff <(echo "${expected_runafter}") <(echo "${actual_runafter}")); then
```
1. With my suggestion below it will become unnecessary to capture diff in a variable. On the opposite, it can be printed as-is.
2. Also, I find it more useful to work with two-column diff and I find the default output hard to comprehend.
3. After that, it may be appropriate to tell what we're checking.

```suggestion
    echo "Checking .tekton/operator-bundle-pipeline.yaml that create-acs-style-snapshot task's runAfter contents match the expected ones (left - expected, right - actual):"
    # Please note that diff won't be printed if there's no diff, so the above message surely needs tweaking.
    if ! diff --side-by-side <(echo "${expected_runafter}") <(echo "${actual_runafter}"); then
```

msugakov reviewed (commented):
Some notes about the validation script.

tommartensen commented on scripts/ci/jobs/check-konflux-pipelines.sh:
```diff
@@ -0,0 +1,27 @@
+    fi
+}
+
+echo "Ensure that modifications to our Konflux pipelines follow our expectations and conventions"
```
Updated the validation script, see run https://github.com/stackrox/stackrox/actions/runs/12579762565/job/35060500687#step:10:1

-> the diff output is shown even if there is no diff. I think that is fine (It didn't show up on my machine)

https://github.com/stackrox/stackrox/pull/13577/commits/07c045ecd8b3e00037b78c4f0ed048073cdc6529

tommartensen commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,71 @@
+
+      .konflux/scripts/render_snapshot.py
+
+  - name: create-snapshot
```
I chose `create`, because it feels closer to the operation we do on the cluster.

tommartensen commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,71 @@
+    env:
+      # For these vars, args: will not work
+      - name: APPLICATION
+        valueFrom:
```
#konflux-users pointed me towards https://konflux-ci.dev/docs/how-tos/testing/integration/creating/#utilizing-the-labels-and-annotations-within-the-integration-pipeline-run

https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1733927292736509?thread_ts=1733920865.416809&cid=C04PZ7H0VA8

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,103 @@
+        "metadata": {
+            "name": snapshot_name,
+            "namespace": namespace,
+            "labels": {
+                "appstudio.openshift.io/build-pipelinerun": pipeline_run_name
+            }
```
Yes.

tommartensen replied:
https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/snapshots/acs-20250102t130214z/overview 

shows the components contained in the snapshot and the PipelineRuns associated with the snapshot (operator-bundle and the integration tests.

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,103 @@
+    ), "Component must have component name, ref, revision and repository set. Check container image labels."
+
+
+def determine_component_version_suffix(application):
```
Changed to `determine_product_version_suffix`. (consistently calling the suffix`product_version` now.

msugakov commented on .github/workflows/style.yaml:
```diff
@@ -75,6 +75,9 @@ jobs:
     - name: Check Policies
       run: scripts/ci/jobs/policy-checks.sh
 
+    - name: Check Konflux pipeline modifications
```
Let's simplify it, please:
```suggestion
    - name: Check Konflux pipelines
```

msugakov commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,100 @@
+        f.write(snapshot["metadata"]["name"])
+
+
+if __name__ == '__main__':
```
If you'll be making more passes over this file and would find the following idea interesting...

Uncle Bob Martin in his explanations of literate programming (in one of [these](https://www.youtube.com/watch?v=7EmboKQH8lM&list=PLmmYSbUCWJ4x1GO839azG_BBw8rkh-zOj) sessions) suggests that the order of functions in the file should be the top-level ones first, their dependencies after, sub-dependencies after and so on.
There's no strict rule and one is impossible to create, because it's all based on human judgment, but the idea is that when you open the file you first see the main thing it does. If you need to learn more details, you start scrolling. Once you got info at the desired depth, you may stop scrolling which is the advantage of such code file organization.

Here, for example, a new `def main():` would go to the top (still `if __name__ == '__main__':` calling `main()` should remain at the bottom by Python users convention), the functions `main()` directly calls would follow after it (probably in the order they are called).
I place the second-level functions (not directly called by `main()` but called from the first-level ones which _are_ called from `main()`) close to where they are called from, obviously under their callers. This way you don't need to find something like `validate_component` at the end of the file but it would rather be close to `process_component`.

E.g. the order of functions in this file could be:
```
main
parse_components_input
determine_product_version_suffix
determine_snapshot_name
process_component
validate_component
construct_snapshot
write_snapshot
```

Up to you.

tommartensen replied:
Moved `validate_component` so now we have: 

```
parse_components_input 
determine_resource_version_suffix 
determine_snapshot_name 
process_component 
validate_component 
construct_snapshot 
write_snapshot
```

msugakov commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -0,0 +1,52 @@
+    description: Image reference.
+    type: string
+  results:
+  - name: GIT_REF
+    description: Git commit for the image's source code if provided as `vcs-ref` image label.
+  - name: GIT_REPO
+    description: Git repository URL for the image's source code if provided as `source-location` image label.
```
Since these became non-optional, please update their descriptions
```suggestion
  - name: GIT_REF
    description: Git commit for the image's source code from `vcs-ref` image label.
  - name: GIT_REPO
    description: Git repository URL for the image's source code from `source-location` image label.
```

msugakov commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,61 @@
+spec:
+  description: Creates a snapshot based on the component names, image refs and Git information passed to the task.
+  params:
+  - name: SOURCE_ARTIFACT
+    description: The Trusted Artifact URI pointing to the application source code.
+    type: string
```
I feel like I may forget it in the future so I'll add a comment now. Once you'll inline `render_snapshot.py`, it seems it won't be necessary to use the git repo at all. Let's not forget to delete this `SOURCE_ARTIFACT` and the corresponding `use-trusted-artifact` step.

msugakov reviewed (commented):
Made a pass. Could still miss some things, will try to catch in the next one(s).

msugakov commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,71 @@
+      - name: PIPELINE_RUN_NAME
+        valueFrom:
+          fieldRef:
+            fieldPath: metadata.name
```
Following up in https://github.com/stackrox/stackrox/pull/13577#discussion_r1901088804
Resolving this comment.

tommartensen commented on .tekton/create-snapshot-task.yaml:
```diff
@@ -0,0 +1,61 @@
+    - name: PIPELINE_RUN_NAME
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.labels['tekton.dev/pipeline']
```
FTR https://tekton.dev/docs/pipelines/labels/#automatic-labeling

tommartensen commented on .tekton/wait-for-image-task.yaml:
```diff
@@ -0,0 +1,52 @@
+
+      # Turning raw_info into an array for easier handling.
+      infos=( $raw_info )
+      if [ ${#infos[@]} -ne 3 ]; then
```
```bash
#!/bin/bash

raw_info="123 546"

# Turning raw_info into an array for easier handling.
infos=( $raw_info )
if [[ "${#infos[@]}" -ne 3 ]]; then
  echo "ERROR: Not all required information was found. Verify that the 'vcs-ref' and 'source-location' labels are set on the image IMAGE."
  exit 1
fi

echo "${infos[0]}"
echo "${infos[1]}"
echo "${infos[2]}"
```

Verified with this script that your suggestion works.

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,100 @@
+    product_version = os.environ["PRODUCT_VERSION"]
+    snapshot_name_result_path = os.environ["SNAPSHOT_NAME_RESULT_PATH"]
+
+    product_version_suffix = determine_product_version_suffix(application)
```
Changed to resource_version_suffix - that also resolves https://github.com/stackrox/stackrox/pull/13577#discussion_r1901139615

tommartensen commented on .konflux/scripts/render_snapshot.py:
```diff
@@ -0,0 +1,100 @@
+    snapshot_name_result_path = os.environ["SNAPSHOT_NAME_RESULT_PATH"]
+
+    product_version_suffix = determine_product_version_suffix(application)
+    snapshot_name = determine_snapshot_name(application, product_version_suffix)
```
Yes, and `product_version_suffix ` (now `resource_version_suffix` - including the `-`) is correct.

msugakov replied:
Following up in https://github.com/stackrox/konflux-tasks/pull/22#discussion_r1903885635

msugakov commented on scripts/ci/jobs/check-konflux-pipelines.sh:
```diff
@@ -6,18 +6,20 @@
+    actual_runafter="$(yq eval '.spec.tasks[] | select(.name == '\"${task_name}\"') | .runAfter[]' "${pipeline_path}")"
 
-    echo "➤ .tekton/operator-bundle-pipeline.yaml // create-acs-style-snapshot: task's runAfter contents shall match the expected ones (left - expected, right - actual)."
+    echo "➤ ${pipeline_path} // checking create-acs-style-snapshot: task's runAfter contents shall match the expected ones (left - expected, right - actual)."
```
```suggestion
    echo "➤ ${pipeline_path} // checking ${task_name}: task's runAfter contents shall match the expected ones (left - expected, right - actual)."
```

msugakov commented on scripts/ci/jobs/check-konflux-pipelines.sh:
```diff
@@ -6,18 +6,20 @@
-    echo "➤ .tekton/operator-bundle-pipeline.yaml // create-acs-style-snapshot: task's runAfter contents shall match the expected ones (left - expected, right - actual)."
+    echo "➤ ${pipeline_path} // checking create-acs-style-snapshot: task's runAfter contents shall match the expected ones (left - expected, right - actual)."
     if ! diff --side-by-side <(echo "${expected_runafter}") <(echo "${actual_runafter}"); then
         echo >&2 -e """
```
(I saw this before but did not comment.)
I don't think that `bash` has any special triple-quote thing to enable multiline strings like Python does. Folks on the internet say that
```bash
a="""
something
"""
```
would be just three strings concatenated together. The first and the third ones are empty.

Strings in bash are allowed to be multiline so you can write
```bash
a="
something
"
```
to the same effect.

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -660,3 +663,129 @@ spec:
+            "revision": "$(tasks.wait-for-scanner-v4-image.results.GIT_REF)"
+          }
+        ]
+    - name: PRODUCT_VERSION
+      value: $(tasks.determine-main-image-tag.results.IMAGE_TAG)
```
Do you mind moving `PRODUCT_VERSION` before `COMPONENTS`?
When there's a large argument like `COMPONENTS`, I'm used to find smaller ones to be provided before it so that when reading the code I can pick up quickly what's the idea of the rest-of parameters without having to scroll to the end. Not sure I could explain it well but if you do the same, you could relate.

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -660,3 +663,129 @@ spec:
       resolver: bundles
+
+  - name: create-acs-style-snapshot
+    # Explicitly running after all other tasks to ensure that
+    # - there are no failures that should prevent a release of the operator-bundle image (missing RPMs signatures, deprecated base images, ...)
+    # - the source image is present as it is required by EC
+    runAfter:
```
I'd suggest you mention in this comment that `scripts/ci/jobs/check-konflux-pipelines.sh` as an easy way to keep `runAfter` list up-to-date.

tommartensen commented:
/test operator-on-push

tommartensen commented:
/test roxctl-on-push

tommartensen commented:
/test scanner-v4-on-push

tommartensen commented:
/test operator-bundle-on-push

tommartensen commented:
/test central-db-on-push operator-bundle-on-push

tommartensen commented:
@msugakov addressed all changes here and in `konflux-tasks`. With the latest task bundle from that PR
* operator-bundle pipeline: https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/pipelineruns/operator-bundle-on-push-xg8zs
* snapshot https://console.redhat.com/application-pipeline/workspaces/rh-acs/applications/acs/snapshots/acs-4-7-x-432-g1193ed49ae-fast-20250106t161152z (EC are failing because of the usual)

msugakov commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -660,3 +663,130 @@ spec:
         value: task
       resolver: bundles
+
+  - name: create-acs-style-snapshot
```
I just realized that it would be nice to put a comment saying that this `create-acs-style-snapshot` is the thing which creates a Snapshot representing a valid set of ACS containers unlike the Snapshots created by Konflux automatically.

This could be a comment but it seems that the Tekton spec allows the `description:` attribute here.
```suggestion
  - name: create-acs-style-snapshot
    description: see my comment
```

tommartensen replied:
done.

msugakov reviewed (approved):
Looks good with one cosmetic comment to address.

Please don't forget to:

- [ ] Update the task bundle to `latest` before merging.
- [ ] Make sure the `Check Konflux pipelines` is required, i.e. its failure prevents merging a PR.

tommartensen commented on .tekton/operator-bundle-pipeline.yaml:
```diff
@@ -660,3 +663,130 @@ spec:
+    params:
+    - name: PRODUCT_VERSION
+      value: $(tasks.determine-main-image-tag.results.IMAGE_TAG)
+    - name: COMPONENTS
+      value: |
+        [
+          {
+            "name": "central-db",
```
I will leave ROX-23316 open and implement this check in another PR.

tommartensen commented:
/test scanner-v4-db-on-push

tommartensen commented:
/test operator-on-push

tommartensen commented:
/test operator-on-push

tommartensen commented:
/test roxctl-on-push

tommartensen commented:
/test scanner-v4-db-on-push

tommartensen commented:
/test scanner-v4-on-push

tommartensen commented:
/test operator-on-push

tommartensen commented:
/test scanner-v4-on-push

tommartensen commented:
/test scanner-v4-db-on-push

tommartensen commented:
/test scanner-v4-on-push
